File: Go/Stack/Product-Of-Array-Except-Self.go
package main

func main() {

}

func productExceptSelf(nums []int) []int {
	result := make([]int, len(nums), len(nums))

	for i := range result {
		result[i] = 1
	}

	prefix := 1
	for i, num := range nums {
		result[i] = prefix
		prefix *= num
	}

	postfix := 1

	for i := len(nums) - 1; i >= 0; i-- {
		result[i] *= postfix
		postfix *= nums[i]
	}

	return result
}

File: Rust/Stack/Product-Of-Array-Except-Self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut result = vec![1; nums.len()];

        let mut prefix: i32 = 1;

        for i in 0..result.len() {
            result[i] = prefix;
            prefix *= nums[i];
        }

        let mut postfix = 1;

        for i in (0..result.len()).rev() {
            result[i] *= postfix;
            postfix *= nums[i];
        }

        result
    }
}
